### PR TITLE
feat: replace `DependsOn` to connector by its generated logical IDs

### PR DESCRIPTION
**Pull request by hoffa**
### Issue #, if available

https://github.com/aws/serverless-application-model/issues/2534

### Description of changes

A connector logical ID in `DependsOn` will be automatically replaced by the logical IDs of its generated resources. This makes situations where `DependsOn` is necessary a little easier (e.g. event source mapping with DLQ), and it's consistent with how currently the abstract resources are referenceable.

For example, the following:

```yaml
DependsOn:
  - SomeConnector
```

Will be replaced by:

```yaml
DependsOn:
  - SomeConnectorWriteLambdaPermission 
```

### Checklist

- [x] Add/update [unit tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions) using:
    - [x] Correct values
    - [ ] Bad/wrong values (None, empty, wrong type, length, etc.)
    - [ ] [Intrinsic Functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html)
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)
- [ ] Update documentation
- [ ] Verify transformed template deploys and application functions as expected
- [ ] Do these changes include any template validations?
    - [ ] Did the newly validated properties support intrinsics prior to adding the validations? (If unsure, please review [Intrinsic Functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html) before proceeding).
        - [ ] Does the pull request ensure that intrinsics remain functional with the new validations?

### Examples?

Please reach out in the comments, if you want to add an example. Examples will be 
added to `sam init` through https://github.com/awslabs/aws-sam-cli-app-templates/

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/translator/input/connector_bucket_to_function.yaml [hunk 1]
```diff
@@ -12,7 +12,7 @@ Resources:
 
   Bucket:
     # See also https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket-notificationconfig.html
-    DependsOn: [MyConnectorWriteLambdaPermission]
+    DependsOn: MyConnector
```
Comment (GavinZZ) [inline tests/translator/input/connector_bucket_to_function.yaml:15]: nit: This should still be [MyConnector] right?

**Review by GavinZZ — APPROVED** [review]
Nice!

File: tests/translator/input/connector_bucket_to_function.yaml (hunk 1 above)
Comment (hoffa) [reply tests/translator/input/connector_bucket_to_function.yaml:15]: Not necessarily, both are correct: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-dependson.html

File: samtranslator/utils/utils.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+def insert_unique(xs: Any, vs: Any) -> List[Any]:
+    """
+    Return copy of `xs` extended with values of `vs` that do not exist in `xs`.
+
+    Inputs are converted to lists if they already aren't.
+    """
+    xs = as_array(copy.deepcopy(xs))
+    vs = as_array(copy.deepcopy(vs))
+
+    for v in vs:
```
Comment (aaythapa) [inline samtranslator/utils/utils.py:19]: Maybe I'm misreading the logic but can't you convert to sets and use the .union function

**Review by awood45 — APPROVED** [review]

File: samtranslator/utils/utils.py (hunk 2 above)
Comment (hoffa) [reply samtranslator/utils/utils.py:19]: Could, but we want it to be always predictable and deterministic (ensures we don't cause changes/redeployments and tests keep working regardless of Python version, and we can determine order without knowing implementation details).

For example:

```python
>>> list(set(['a', 'wat', 'lol']))
['lol', 'wat', 'a']
```

The order changes due to implementation details (the hashing); we don't want to rely on it staying as-is (in fact during the Python 2 transition to Python 3, there was a lot of work needed to keep code that relied on the underlying hashing algorithm backwards compatible).

Comment (aaythapa) [reply samtranslator/utils/utils.py:19]: Ah I see, I thought .union had a guaranteed order. Thanks for the clarification

**Review by hawflau — APPROVED** [review]
LGTM. I originally thought this is better done via a plugin. But this seems more precise as each connector handles its own logical ID.

**Pull request merged**